src/goto-instrument/cover.cpp

```diff
@@ -266,32 +226,108 @@ bool instrument_cover_goals(
     function_filters.add(
       util_make_unique<trivial_functions_filtert>(message_handler));
 
-  msg.status() << "Instrumenting coverage goals" << messaget::eom;
-
-  instrument_cover_goals(
-    goto_functions, instrumenters, function_filters, message_handler);
+  config->traces_must_terminate =
+    options.get_bool_option("cover-traces-must-terminate");
 
-  function_filters.report_anomalies();
-  goal_filters.report_anomalies();
+  return config;
+}
 
-  if(options.get_bool_option("cover-traces-must-terminate"))
+/// Instruments a single goto program based on the given configuration
+/// \param config: configuration, produced using get_cover_config
+/// \param function_id: function name
+/// \param function: function function to instrument
+/// \param message_handler: log output
+static void instrument_cover_goals(
+  const cover_configt &config,
+  const irep_idt &function_id,
+  goto_functionst::goto_functiont &function,
+  message_handlert &message_handler)
+{
+  if(!config.keep_assertions)
   {
-    // instrument an additional goal in CPROVER_START. This will rephrase
-    // the reachability problem  by asking BMC to provide a solution that
-    // satisfies a goal while getting to the end of the program-under-test.
-    const auto sf_it=
-      goto_functions.function_map.find(goto_functions.entry_point());
-    if(sf_it==goto_functions.function_map.end())
+    Forall_goto_program_instructions(i_it, function.body)
```

Comment:
Can we here swap to using range based for loop? Or perhaps even a `std::foreach` since the body is so simple

Reply:
I'm going to plead "moved code" plus slim benefit and leave this alone :)

Reply:
I argue that the existing coverage tests will do this for us when this comes into use. Those tests will appear with the symex-driven lazy loading PR, which will be this code's first user and will introduce a variant of the cbmc-java test directory amongst others, running all tests with symex-driven lazy loading (and therefore per-function GOTO program passes) enabled.